library/std/src/sys/pal/windows/api.rs

```diff
@@ -34,6 +34,105 @@ use core::ptr::addr_of;
 
 use super::c;
 
+/// Creates a null-terminated UTF-16 string from a str.
+macro_rules! wide_str {
+    ($str:literal) => {{
+        const _: () = {
+            if core::slice::memchr::memchr(0, $str.as_bytes()).is_some() {
+                panic!("null terminated strings cannot contain interior nulls");
+            }
+        };
+        utf16!(concat!($str, '\0'))
+    }};
+}
+
+/// Creates a UTF-16 string from a str without null termination.
+macro_rules! utf16 {
+    // Note: this macro uses triple underscores to avoid const cycles
+    ($str:expr) => {{
+        const ___UTF8: &str = $str;
+        const ___UTF16_LEN: usize = crate::sys::pal::windows::api::utf16_len(___UTF8);
+        const ___UTF16: [u16; ___UTF16_LEN] = crate::sys::pal::windows::api::to_utf16(___UTF8);
+        &___UTF16
+    }};
+}
+
+#[cfg(test)]
+mod tests;
+
+/// Gets the UTF-16 length of a UTF-8 string, for use in the wide_str macro.
+pub const fn utf16_len(s: &str) -> usize {
+    let s = s.as_bytes();
+    let mut i = 0;
+    let mut len = 0;
+    while i < s.len() {
+        // the length of a UTF-8 encoded code-point is given by the number of
+        // leading ones, except in the case of ASCII.
+        let utf8_len = match s[i].leading_ones() {
+            0 => 1,
+            n => n as usize,
+        };
+        i += utf8_len;
+        len += if utf8_len < 4 { 1 } else { 2 };
```

Comment:
hmm. this feels slightly wrong. the 3-byte case...
1110xxxx 10xxxxxx 10xxxxxx?

that's 16 bits of raw entropy, but if memory serves, not all u16s are individually valid in UTF-16? because of the surrogate reservation? so there are some 3-byte UTF-8 sequences that are two u16s, aren't there? there's some more nuances of the encoding that I'm not quite remembering, and I know not all 3-byte sequences are valid either, but...

if I'm wrong, this needs a comment explaining why I'm wrong, since it seems like a "naive" bit of code, and a sophomoric programmer might try to "correct" it otherwise.

Reply:
Those are also not valid UTF-8. I'll add a comment. See https://en.wikipedia.org/wiki/UTF-8#Invalid_sequences_and_error_handling

> Since RFC 3629 (November 2003), the high and low surrogate halves used by [UTF-16](https://en.wikipedia.org/wiki/UTF-16) (U+D800 through U+DFFF) and code points not encodable by UTF-16 (those after U+10FFFF) are not legal Unicode values, and their UTF-8 encoding must be treated as an invalid byte sequence. Not decoding unpaired surrogate halves makes it impossible to store invalid UTF-16 (such as Windows filenames or UTF-16 that has been split between the surrogates) as UTF-8,[[e]](https://en.wikipedia.org/wiki/UTF-8#cite_note-PEP_529_quote-33) while it is possible with [WTF-8](https://en.wikipedia.org/wiki/UTF-8#WTF-8).

Reply:
I'll add a comment. UTF-16 can encode values U+0000 to U+D7FF and U+E000 to U+FFFF in [a single u16](https://en.wikipedia.org/wiki/UTF-16#U+0000_to_U+D7FF_and_U+E000_to_U+FFFF).

The hole in the middle are for the surrogate pairs but these [aren't valid UTF-8](https://en.wikipedia.org/wiki/UTF-8#Invalid_sequences_and_error_handling) either:

> Since RFC 3629 (November 2003), the high and low surrogate halves used by [UTF-16](https://en.wikipedia.org/wiki/UTF-16) (U+D800 through U+DFFF) and code points not encodable by UTF-16 (those after U+10FFFF) are not legal Unicode values, and their UTF-8 encoding must be treated as an invalid byte sequence. Not decoding unpaired surrogate halves makes it impossible to store invalid UTF-16 (such as Windows filenames or UTF-16 that has been split between the surrogates) as UTF-8,[[e]](https://en.wikipedia.org/wiki/UTF-8#cite_note-PEP_529_quote-33) while it is possible with [WTF-8](https://en.wikipedia.org/wiki/UTF-8#WTF-8).

Reply:
...The answer to this bugs me on some level because the encoding could be more compact, I think, even in the presence of the surrogate pair codepoints, by offsetting the encoding ranges by the fixed number of the last range, even without any further cleverness.

Reply:
Time to propose a new Unicode encoding?

Reply:
Ah, so now I understand. This problem is the problem of overlong encodings. They were once accepted, so we have a large number of unused bits.